### PR TITLE
[SPARK-18658][SQL] Write text records directly to a FileOutputStream

### PR DESCRIPTION
**Pull request by NathanHowell**
## What changes were proposed in this pull request?

This replaces uses of `TextOutputFormat` with an `OutputStream`, which will either write directly to the filesystem or indirectly via a compressor (if so configured). This avoids intermediate buffering.

The inverse of this (reading directly from a stream) is necessary for streaming large JSON records (when `wholeFile` is enabled) so I wanted to keep the read and write paths symmetric.

## How was this patch tested?

Existing unit tests.

### CODE REVIEW

**Comment by NathanHowell** [comment]
This touches a fair number of components. I also haven't done any performance testing to see what the impact of this is. Curious what your thoughts are?

cc/ @marmbrus @rxin @JoshRosen

**Comment by rxin** [comment]
Does this work against file systems with HDFS API (not local posix)? If yes, sounds good!

**Comment by NathanHowell** [comment]
Yep. It uses the Hadoop `FileSystem` class to open files, just like `TextOutputFormat` does.

**Comment by rxin** [comment]
Yea then this is definitely fine.

**Comment by JoshRosen** [comment]
Jenkins, this is ok to test

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonGenerator.scala [hunk 1]
```diff
@@ -194,4 +194,8 @@ private[sql] class JacksonGenerator(
       writeFields(row, schema, rootFieldWriters)
     }
   }
+
+  private[sql] def writeLineEnding(): Unit = {
+    gen.writeRaw('\n')
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonGenerator.scala]: Hm, is this safe to guess the newline is always `\n`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/text/TextFileFormat.scala [hunk 2]
```diff
@@ -132,39 +128,17 @@ class TextOutputWriter(
...
+  private val writer = CodecStreams.getOutputStream(context, new Path(path))
 
   override def write(row: Row): Unit = throw new UnsupportedOperationException("call writeInternal")
 
   override protected[sql] def writeInternal(row: InternalRow): Unit = {
     val utf8string = row.getUTF8String(0)
-    buffer.set(utf8string.getBytes)
-    recordWriter.write(NullWritable.get(), buffer)
+    writer.write(utf8string.getBytes)
+    writer.write('\n')
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/text/TextFileFormat.scala]: Here too and for CSV as well.

Comment (NathanHowell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/text/TextFileFormat.scala]: This mirrors what Hadoop code does, see https://github.com/apache/hadoop/blob/f67237cbe7bc48a1b9088e990800b37529f1db2a/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-core/src/main/java/org/apache/hadoop/mapreduce/lib/output/TextOutputFormat.java#L48-L49

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/CodecStreams.scala [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+import java.io.{OutputStream, OutputStreamWriter}
+import java.nio.charset.{Charset, StandardCharsets}
+
+import org.apache.hadoop.fs.Path
+import org.apache.hadoop.io.compress._
+import org.apache.hadoop.mapreduce.JobContext
+import org.apache.hadoop.mapreduce.lib.output.FileOutputFormat
+import org.apache.hadoop.util.ReflectionUtils
+
+private[spark] object CodecStreams {
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/CodecStreams.scala]: I guess we don't need `private[spark]` here assuming from the commit here (https://github.com/apache/spark/commit/511f52f8423e151b0d0133baf040d34a0af3d422)

**Comment by NathanHowell** [comment]
Doh, forgot to run the Hive tests. Should be fixed now.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/CodecStreams.scala (hunk 3 above)
Comment (NathanHowell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/CodecStreams.scala]: Looks that way, I've removed it.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/CodecStreams.scala [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+      Some(ReflectionUtils.newInstance(compressorClass, context.getConfiguration))
+    } else {
+      file.flatMap { path =>
+        val compressionCodecs = new CompressionCodecFactory(context.getConfiguration)
+        Option(compressionCodecs.getCodec(path))
+      }
+    }
+  }
+
+  /** Create a new file and open it for writing.
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/CodecStreams.scala]: can you fix the comment style here? we don't use scaladoc style comment in Spark.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonGenerator.scala [hunk 5]
```diff
@@ -194,4 +194,8 @@ private[sql] class JacksonGenerator(
       writeFields(row, schema, rootFieldWriters)
     }
   }
+
+  private[sql] def writeLineEnding(): Unit = {
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonGenerator.scala]: remove private[sql] here to be consistent with other methods

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/CodecStreams.scala [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+   * If compression is enabled in the [[JobContext]] the stream will write compressed data to disk.
+   * An exception will be thrown if the file already exists.
+   */
+  def getOutputStream(context: JobContext, file: Path): OutputStream = {
+    val fs = file.getFileSystem(context.getConfiguration)
+    val outputStream: OutputStream = fs.create(file, false)
+
+    getCompressionCodec(context, Some(file)).fold(outputStream) { codec =>
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/CodecStreams.scala]: I know you like Haskell but every time I see a fold I have to spend an extra 5 secs checking what it is doing :) Can we simplify this one?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/CodecStreams.scala [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+        Option(compressionCodecs.getCodec(path))
+      }
+    }
+  }
+
+  /** Create a new file and open it for writing.
+   * If compression is enabled in the [[JobContext]] the stream will write compressed data to disk.
+   * An exception will be thrown if the file already exists.
+   */
+  def getOutputStream(context: JobContext, file: Path): OutputStream = {
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/CodecStreams.scala]: getOutputStream -> createOutputStream

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/CodecStreams.scala [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+  def getOutputStreamWriter(
+      context: JobContext,
+      file: Path,
+      charset: Charset = StandardCharsets.UTF_8): OutputStreamWriter = {
+    new OutputStreamWriter(getOutputStream(context, file), charset)
+  }
+
+  /** Returns the compression codec extension to be used in a file name, e.g. ".gzip"). */
+  def getCompressionExtension(context: JobContext): String = {
+    getCompressionCodec(context).fold("") { code =>
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/CodecStreams.scala]: maybe

```
getCompressionCodec(context).map(_. getDefaultExtension).getOrElse("")
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/text/TextFileFormat.scala [hunk 9]
```diff
@@ -132,39 +128,17 @@ class TextOutputWriter(
...
-  }
+  private val writer = CodecStreams.getOutputStream(context, new Path(path))
 
   override def write(row: Row): Unit = throw new UnsupportedOperationException("call writeInternal")
 
   override protected[sql] def writeInternal(row: InternalRow): Unit = {
     val utf8string = row.getUTF8String(0)
-    buffer.set(utf8string.getBytes)
-    recordWriter.write(NullWritable.get(), buffer)
+    writer.write(utf8string.getBytes)
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/text/TextFileFormat.scala]: can you check UTF8String's implementation to make sure we are not creating a new array for each row?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVRelation.scala [hunk 10]
```diff
@@ -245,24 +230,12 @@ private[csv] class CsvOutputWriter(
...
-    }
-  }
-
-  private def flush(): Unit = {
-    val lines = csvWriter.flush()
-    if (lines.nonEmpty) {
-      text.set(lines)
-      recordWriter.write(NullWritable.get(), text)
-    }
+    csvWriter.writeRow(rowToString(row), printHeader)
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVRelation.scala]: we can probably optimize this as well - but not a huge deal in the first pr.

**Comment by srowen** [comment]
I was going to say, hm, are we sure we want to reimplement / go around the Hadoop support for this? but in practice it looks like it actually simplifies some things. At the moment I can't think of any particular behaviors we're missing by avoiding the Input/OutputFormat.

But CC @vanzin @steveloughran for any comment.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/CodecStreams.scala (hunk 6 above)
Comment (NathanHowell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/CodecStreams.scala]: Yah, it's a terrible name (and it's not a fold). I'll replace them with `.map(...).getOrElse`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/text/TextFileFormat.scala (hunk 9 above)
Comment (NathanHowell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/text/TextFileFormat.scala]: It is creating a new array, I'll pass the internal one through instead.

Comment (NathanHowell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/text/TextFileFormat.scala]: Done, but I'm not 100% sure about the implementation. Can you have someone more familiar with `UTF8String`'s internals double check it?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/CodecStreams.scala [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+        val compressionCodecs = new CompressionCodecFactory(context.getConfiguration)
+        Option(compressionCodecs.getCodec(path))
+      }
+    }
+  }
+
+  /**
+   * Create a new file and open it for writing.
+   * If compression is enabled in the [[JobContext]] the stream will write compressed data to disk.
+   * An exception will be thrown if the file already exists.
```
Comment (steveloughran) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/CodecStreams.scala]: will "probably" be thrown; object stores have issues there

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 12]
```diff
@@ -147,6 +147,17 @@ public void writeTo(ByteBuffer buffer) {
     buffer.position(pos + numBytes);
   }
 
+  public void writeTo(OutputStream out) throws IOException {
```
Comment (NathanHowell) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: Agreed.

**Comment by steveloughran** [comment]
AFAIK, the big thing the FileOutputFormat really adds is not the compression, but the output committer and the stuff to go with that (working directories, paths, etc etc). If you aren't going near that, and just want a fast write of .csv and jackson with optional compression, well, I don't see anything in the code I'd run away from.

If you do want to think about how to write CSV files during the output of speculative work in the presence of failures, well, that's where the mapred.lib.output code really comes out to play.

Otherwise, in general PR review mode: Tests? What if the code asks for a committer that isn't there, passes in null sequences in rows to write, tries to hit the buffer corner cases. Hopefully those exist already, but if not, now is a good time to try to break things.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonGenerator.scala [hunk 13]
```diff
@@ -194,4 +194,8 @@ private[sql] class JacksonGenerator(
       writeFields(row, schema, rootFieldWriters)
     }
   }
+
+  def writeLineEnding(): Unit = {
+    gen.writeRaw('\n')
```
Comment (steveloughran) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonGenerator.scala]: Presumably just a convoluted way to create a byte array containing the byte `0x0a` then

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/CodecStreams.scala (hunk 11 above)
Comment (NathanHowell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/CodecStreams.scala]: Is this a problem with Hadoop in general? The FileSystem docs also specify this behavior:

```
  /**
   * Create an FSDataOutputStream at the indicated Path.
   * @param f the file to create
   * @param overwrite if a file with this name already exists, then if true,
   *   the file will be overwritten, and if false an exception will be thrown.
   */
```

**Comment by NathanHowell** [comment]
@steveloughran Spark is handling the output committing somewhere further up the stack. The path being passed in to `OutputWriterFactory.newInstance` is to a temporary file, such as `/private/var/folders/sq/vmncyd7506q_ch43llrwr8sn6zfknl/T/spark-3db2844b-1f3c-45c2-8bf4-8a3c81440e38/_temporary/0/_temporary/attempt_20161201081833_0000_m_000000_0/part-00000-8dd44cea-c01e-4bfe-ab03-641ebce18afb.txt`.

I'll make a pass through the existing tests to see if anything obvious is missing.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonGenerator.scala (hunk 13 above)
Comment (NathanHowell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonGenerator.scala]: That is my assumption. I'm also assuming that writing a single byte is slightly more efficient than writing an array of a single byte.

**Comment by steveloughran** [comment]
I ask about committers as I'm staring at the V1 and V2 committer APIs right now related to S3 destinations; not directly related to this though.

**Comment by rxin** [comment]
@srowen yea the hadoop format api is pretty awkward to use, and actually makes everything more complicated than needed.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVRelation.scala (hunk 10 above)
Comment (NathanHowell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVRelation.scala]: The uniVocity CSV writer converts every column to a `String` before writing so it's (probably?) not possible to further optimize this without doing a whole bunch of work. I only did a quick scan through their code though.

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java (hunk 12 above)
Comment (NathanHowell) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: I've added a few tests for this method.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVRelation.scala (hunk 10 above)
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVRelation.scala]: Yup I asked this question but never had time to follow-up: https://github.com/uniVocity/univocity-parsers/issues/99

**Comment by rxin** [comment]
Merging in master. Thanks.

**Pull request closed without merging**